ログインページの追加

rakushoo opened the pull request:
# ログイン機能を追加
udemyのうち、signup, login, logout の実装をしました。

# 確認手順
## 準備（DBの用意）
% docker compose up -d
% cd backend
% npx prisma migrate dev
% npx prisma generate

## backend立ち上げ
% cd ../backend
% yarn install && yarn build && yarn start:dev

## frontend立ち上げ
% cd ../frontend
% yarn install && yarn build && yarn run dev


## 後処理（DBコンテナの削除）
% docker compose rm -s -f -v

# 確認してもらいたい内容
* http://localhost:3000/auth_index にアクセス。
* 初めての場合は「Don't have an account? Register」をクリックするとボタンがRegisterに変わるので、EmailとPasswordを設定。自動でログインします。
* |-> のリンクをクリックするとログアウトします。

rakushoo committed:
test commit

rakushoo committed:
add: auth_indexを追加

SignUp, LogIn, LogOut を実装

rakushoo commented:
### 現在の課題
frontendでyarn buildを実行した際、下記エラーによりlingに失敗してしまいます。yarn run dev は実行できるので、動作は確認できます。
Type error: Cannot find type definition file for 'yup'.

package.jsonにはyupは含まれているのですが、yarn installだけでは定義が見つからない、となってしまいます。
調査、対応方法のヒントなどあれば教えていただきたいです。

takashi247 commented on frontend/.env:
```diff
@@ -0,0 +1,8 @@
+# Environment variables declared in this file are automatically made available to Prisma.
```
.env.localが.gitignoreに引っかかってプッシュされていないので、.envファイルの中に下記を追加しないとちゃんと動きませんでした。

```
NEXT_PUBLIC_API_URL="http://localhost:3001"
```

rakushoo replied:
ご確認ありがとうございました！
yupの修正と共に.envの方へ追記しておきました。

rakushoo committed:
fix: build時のlintエラー対応

yupをremove->addして再インストールした

rakushoo committed:
fix: frontendにbackendのURL環境変数を追加

yuta-fujimoto commented on frontend/pages/auth_index.tsx:
```diff
@@ -0,0 +1,124 @@
+        </Alert>
+      )}
+
+      <form onSubmit={form.onSubmit(handleSubmit as VoidFunction)}>
```
👀

rakushoo replied:
eslintの下記エラー対策で as VoidFunction を追記しました。
動作はできますが、対応方法として合ってるか自信ないです。。
もう少し全体像が把握できてから見直します。

80:37  error  Promise returned in function argument where a void return was expected  @typescript-eslint/no-misused-promises

yuta-fujimoto replied:
そんな型あるの知らなかったなぁと思ってました。いまのところ僕含めて他の実装もそれで放置してる気がします。

takashi247 replied:
この部分、handleSubmit関数を下記の通りに修正すると、`as VoidFunction`をつけなくてもリンターに怒られなくなったので、参考までに共有しておきます。
ただ、結局この対処法も、Promiseが返ってきてるところを全部voidにして、無視してるだけなので、ちょっとasync/await周りの処理については、一旦、全体が形になったタイミングで、まとめて総チェックした方が良さそうですね。

```
const handleSubmit = () => {
    try {
      console.log(process.env.NEXT_PUBLIC_API_URL);
      if (process.env.NEXT_PUBLIC_API_URL) {
        if (isRegister) {
          const url_signup = `${process.env.NEXT_PUBLIC_API_URL}/auth/signup`;
          void axios.post(url_signup, {
            email: form.values.email,
            password: form.values.password,
          });
        }
        void axios.post(`${process.env.NEXT_PUBLIC_API_URL}/auth/login`, {
          email: form.values.email,
          password: form.values.password,
        });
        form.reset();
        void router.push('/');
      }
    } catch (e) {
      if (axios.isAxiosError(e) && e.response && e.response.data) {
        setError(e.message);
      }
    }
  };
```

rakushoo replied:
提案ありがとうございます！
まだ理解が及ばず、asyncとPromiseも理解してから、一気に見直ししたいですね。

yuta-fujimoto commented on backend/src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,63 @@
+export class AuthController {
+  constructor(private readonly authService: AuthService) {}
+
+  @Get('/csrf')
```
/つけたりつけなかったりっていうのは何か理由ありますか？

rakushoo replied:
確かに気になってたのですが、udemyの課題通りで今は理由はわかっていません。
/csrf、/logoutだけあって、signup,logoutには無いので、あまり意味はないのかもですが。
ひとまず写経して、全体見えてから修正していきたいと考えてました。

<参照>
https://github.com/GomaGoma676/restapi-todo-nestjs/blob/main/src/auth/auth.controller.ts

yuta-fujimoto replied:
まぁ動作見てる限り/あってもなくても同じ感じしますけどね。了解しました！またお願いします

yuta-fujimoto commented:
ログインなりJWTなりでで要求されてるものが多くて大変ですね、、
動作自体は確認できました。自分もあんまりよくわかってないですが全部のファイルが必要なんでしたっけ
`frontend/pages/_document.tsx`とかこれは一部ですけど`frontend/types/index.ts`のeditedTaskとか

rakushoo commented:
`frontend/pages/_document.tsx`
こちらは、Mantine UIを使うときの設定で追加したファイルです。
_app.tsx と共に必要なファイルのようです。
https://mantine.dev/guides/next/
ここのチームではMUIを使うとのことだったと思うので、必要に応じて置き換えていければって考えています。

`frontend/types/index.ts`のeditedTask
これは不要でした。udemyのタスクリストの機能は持ってこなかったので不要な型になります。
削除しておきます。確認ありがとうございます！

yuta-fujimoto commented:
なるほど、知らなかったです。あとで対応お願いします
小出しで申し訳ないんですが、`backend/package.json.orig`も必要なければ消しておいてもらえると助かります。~~あとは僕は大丈夫です。~~
すみません、一個だけ質問しました。

yuta-fujimoto commented on backend/src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,63 @@
+
+  @HttpCode(HttpStatus.OK)
+  @Post('login')
+  async login(
```
ここだけasyncつける理由ってパッと思いつかなかったんですが何かありますか？

rakushoo replied:
試しに消したところ、login関数内でawaitを使っている都合上、asyncを付けないと指摘されるため、残しています。
async/awaitはまとめて理解したいですね。

yuta-fujimoto replied:
なるほど、理解しました。ありがとうございます！

rakushoo committed:
Merge branch 'main' into add_simple_login

takashi247 commented on backend/src/user/user.controller.ts:
```diff
@@ -0,0 +1,33 @@
+    const req_user: RequestUser = req.user;
+
+    return this.userService.updateUser(req_user.id, dto);
+    // return this.userService.updateUser(req.user.id, dto);
```
ここは次の更新のタイミングで削除してよさそうですね。

rakushoo replied:
削除しました。

takashi247 commented on frontend/components/Layout.tsx:
```diff
@@ -0,0 +1,20 @@
+import { FC, ReactNode } from 'react';
```
`frontend/components/game/home/Layout.tsx`とほぼ内容が一緒なので、追って、統合したいですね。

takashi247 commented on frontend/pages/index.tsx:
```diff
@@ -16,6 +32,14 @@ const Home: NextPage = () => {
         </Link>
       </Stack>
+      <Layout title="Task Board">
+        <LogoutIcon
```
最終的にはログアウトのボタンもHeaderコンポーネントの中に組み込む感じにするといい感じになりそう。

takashi247 commented on frontend/styles/globals.css:
```diff
@@ -1,16 +1,3 @@
-html,
```
このファイルは以前 @ryo-manba さんが #26 の対応で、バックグラウンドカラーの適用をオフにするための設定とかを修正してくれているところなので、 @rakushoo のパートの脱tailwindができたタイミングで復活させる必要がありそうですね。

ryo-manba replied:
あーこれ、global.cssはそのままで一番上の方にtailwindの設定を書いてもらうだけでも問題はないですね。(多分ヘッダーみたいな扱いだと思うので)
```
@tailwind base;
@tailwind components;
@tailwind utilities;
```

takashi247 replied:
おー、そうなんですね！ありがとうございます！
@rakushoo ここだけでも、マージする前に修正してもらった方がいいかと思います。

ryo-manba replied:
あとbackend/package.json.origも不要だと思われるのでついでに削除していただけると助かります！

rakushoo replied:
global.css に元のコードを復活させ、最上部にtailwindの定義を追加した形にしました。

package.json.orig は削除しました。

takashi247 commented on frontend/types/index.ts:
```diff
@@ -0,0 +1,9 @@
+  email: string;
+  password: string;
+};
+export type EditedTask = {
```
この型はタスク管理アプリ用のものな感じがするので、削除しちゃって問題ないかもですね。

rakushoo replied:
削除しました。

takashi247 reviewed (approved):
大作、お疲れさまでした。いくつかコメントつけさせてもらいましたが、ローカルでの挙動は確認できたので、一旦マージしてしまっていいかなと思います。
tailwindやmantineでの実装になっている部分をMUIベースのUIに切り替えていく中で、修正したり復活させたりする必要があるファイルがありそうなので、その部分を注意して今後の実装を進めていく必要がありそうだなと感じました。

rakushoo committed:
fix: 不要な処理を整理する

package.json.orig: ファイル削除
auth.controller.ts: ルーティングの/スラッシュを削除
user.controller.ts: 不要なコメントの削除
index.ts: 使用しない処理を削除(udemy参考にしたtask関係の定義)
global.css: #26 で対応した元々の定義を復活

yuta-fujimoto reviewed (approved):
対応ありがとうございました！mergeします。
